In [0]:
import keras
keras.__version__

Using TensorFlow backend.


'2.1.6'

# Introduccion a convnets

Primero creamos un modelo simple:

In [0]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

Vemos como va quedando la arquitectura de la convnet:

In [0]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


Agregamos un par de capas mas al final de la red para redimensionar el resultado de las capas convolucionales.

In [0]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

Como vamos a hacer una clasificacion de 10 clases agregamos un layer final con 10 salidas y activacion softmax.

Veamos como queda la red final:


In [0]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                36928     
__________

Arriba se puede ver que las salidas de form `(3, 3, 64)` se aplanaron en vectores de forma `(576,)`, antes de pasar por los layers de tipo `Dense`.

Pasamos a entrenar la red en el dataset MNIST:

In [0]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

11493376/11490434 [==============================] - 1s 0us/step


In [0]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 13s 216us/step - loss: 0.1789 - acc: 0.9451
Epoch 2/5
60000/60000 [==============================] - 12s 193us/step - loss: 0.0476 - acc: 0.9853
Epoch 3/5
40256/60000 [===================>..........] - ETA: 3s - loss: 0.0327 - acc: 0.989760000/60000 [==============================] - 12s 192us/step - loss: 0.0333 - acc: 0.9899
Epoch 4/5
60000/60000 [==============================] - 12s 192us/step - loss: 0.0257 - acc: 0.9919
Epoch 5/5
56320/60000 [===========================>..] - ETA: 0s - loss: 0.0201 - acc: 0.993660000/60000 [==============================] - 12s 197us/step - loss: 0.0199 - acc: 0.9936


Finalmente evaluamos el modelo fiteado.

In [0]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 133us/step


In [0]:
test_acc

0.9926